src/main/kotlin/com/exactpro/th2/http/server/util/ResponseBuilder.kt

```diff
@@ -0,0 +1,131 @@
+        if (message.hasParentEventId()) {
+            eventId = message.parentEventId
+        }
+
+    }
+
+    fun setBody(message: RawMessage) = apply {
+        body = message
+        if (message.hasParentEventId()) {
+            eventId = message.parentEventId
+        }
+    }
+
+    fun setGroup(messages: MessageGroup) = apply {
+        when (messages.messagesCount) {
+            0 -> error("Message group is empty")
+            1 -> messages.getMessages(0).run {
+                when {
+                    hasMessage() -> setHead(message)
+                    hasRawMessage() -> setBody(rawMessage)
+                    else -> error("Single message in group is neither parsed nor raw: ${toPrettyString()}")
+                }
+            }
+            2 -> {
+                setHead(messages.getMessages(0).toParsed("Head"))
+                setBody(messages.getMessages(1).toRaw("Body"))
+            }
+            else -> error("Message group contains more than 2 messages")
+        }
+    }
+
+    fun build(): RawHttpResponse<LinkedData> {
+        metadata.putAll(body.metadata.propertiesMap)
+
+        val code: Int = head.getInt(HEADERS_CODE_FIELD) ?: metadata[CODE_PROPERTY]?.toInt() ?: DEFAULT_CODE
+        val reason = head.getString(HEADERS_REASON_FIELD) ?: metadata[REASON_PROPERTY] ?: DEFAULT_REASON
+        val statusLine = StatusLine(HttpVersion.HTTP_1_1, code, reason)
+        val httpBody = body.body.toByteArray()
+
+        val httpHeaders = RawHttpHeaders.newBuilder()
+        httpHeaders.overwrite("Content-Length", httpBody.size.toString())
+        head.getList(HEADERS_FIELD)?.forEach {
+            require(it.hasMessageValue()) { "Item of '$HEADERS_FIELD' field list is not a message: ${it.toPrettyString()}" }
+            val message = it.messageValue
+            val name = message.getString(HEADER_NAME_FIELD)
+                ?: error("Header message has no $HEADER_NAME_FIELD field: ${message.toPrettyString()}")
+            val value = message.getString(HEADER_VALUE_FIELD)
+                ?: error("Header message has no $HEADER_VALUE_FIELD field: ${message.toPrettyString()}")
+            httpHeaders.overwrite(name, value)
+        }
+        if (httpBody != null && CONTENT_TYPE_HEADER !in httpHeaders.headerNames) {
+            metadata[CONTENT_TYPE_PROPERTY]?.run {
+                split(HEADER_VALUE_SEPARATOR).forEach {
+                    httpHeaders.with(CONTENT_TYPE_HEADER, it.trim())
+                }
+            }
+        }
+        val uuid = head.metadata.propertiesMap["uuid"] ?: body.metadata.propertiesMap["uuid"]
+        checkNotNull(uuid) { "UUID is required" }
+
+        val messagesId = listOf(head.metadata.id, body.metadata.id)
+
+        return RawHttpResponse<LinkedData>(
+            LinkedData(uuid, eventId, messagesId),
+            null,
+            statusLine,
+            httpHeaders.build(),
+            EagerBodyReader(httpBody)
+        )
```

Comment:
Doing conversion in a builder seems a bit out of place. Looks like this builder can be replaced by utility functions